generator/src/main/java/com/graphicsfuzz/generator/mutateapi/NumericValue.java

```diff
@@ -0,0 +1,150 @@
+import com.graphicsfuzz.common.ast.type.Type;
+import com.graphicsfuzz.common.util.IRandom;
+import com.graphicsfuzz.generator.semanticschanging.LiteralFuzzer;
+import java.util.Optional;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+public class NumericValue implements Value {
+
+  private final BasicType basicType;
+  private final Optional<Number> value;
+
+  public NumericValue(BasicType basicType, Optional<Number> value) {
+    this.basicType = basicType;
+    this.value = value;
+  }
+
+  /**
+   * An empty value constructor.
+   *
+   * @param basicType of Value
+   */
+  public NumericValue(BasicType basicType) {
+    this(basicType, Optional.empty());
+  }
+
+
+  @Override
+  public Type getType() {
+    return basicType;
+  }
+
+  @Override
+  public boolean valueIsKnown() {
+    return value.isPresent();
+  }
+
+  @Override
+  public Expr generateLiteral(LiteralFuzzer literalFuzzer) {
+    if (valueIsKnown()) {
+      if (basicType == BasicType.FLOAT) {
+        return new FloatConstantExpr(value.get().toString());
+      }
+      if (basicType == BasicType.INT) {
+        return new IntConstantExpr(value.get().toString());
+      }
+      if (basicType == BasicType.UINT) {
+        return new UIntConstantExpr(value.get().toString() + "u");
+      }
+    }
+    return literalFuzzer.fuzz(basicType).orElse(null);
+  }
+
+  public Optional<Number> getValue() {
+    return value;
+  }
+
+  public Pair<Optional<Number>, Optional<Number>> getPairSum(IRandom generator) {
+
+    if (!valueIsKnown()) {
+      return new ImmutablePair<>(Optional.empty(), Optional.empty());
+    }
+
+    // To find two numbers whose sum is equal to the given value X. Following the
+    // equation X = A + B, we first need to randomly generate a number A which will be used as
+    // the left expression, and subtract it with the original value X. Next, as B = X - A, we use
+    // the outcome of such subtraction as the right expression. Finally, the result of adding two
+    // numbers A and B would be equal to the original value X.
+    //
+    // For example, if a number 5 is an input and we generate a random number 3, we then subtract 5
+    // with 3 which will give 2 as the result. Next we derive left and right expressions from
+    // these two numbers and use them when generating a binary expression.
+
+    if (getType() == BasicType.FLOAT) {
+      float original = value.get().floatValue();
+      float left = generator.nextFloat();
+      float right = original - left;
+      return new ImmutablePair<>(Optional.of(left), Optional.of(right));
+    }
+
+    if (getType() == BasicType.INT) {
+      int original = value.get().intValue();
+      int left = generator.nextInt(original + 1);
+      int right = original - left;
+      return new ImmutablePair<>(Optional.of(left), Optional.of(right));
+    }
+
+    throw new RuntimeException("Should be unreachable");
+  }
+
```

Comment:
Make analogous changes to "equals" as suggested in BooleanValue.

Reply:
And in the equals method, can't we just return:

   this.basicType == thatNumericValue.basicType && this.value.equals(thatNumericValue.value())

?